gen/internals.py

```diff
@@ -184,28 +212,131 @@ def add_entry(self, entry, replace_existing):
         self.add_conditional_scope(entry, [])
 
 
-# NOTE: This exception should never escape the DFSArgumentCalculator
+# NOTE: This exception should never escape the Resolver
 class CalculatorError(Exception):
-    def __init__(self, message, chain=[]):
-        assert isinstance(message, str)
+
+    def __init__(self, message: str, chain=None):
+        if chain is None:
+            chain = list()
         assert isinstance(chain, list)
         self.message = message
         self.chain = chain
         super().__init__(message)
 
 
+class SkipError(Exception):
+    """Raised when this key should silently produce no errors."""
+    pass
+
+
+class Resolvable:
+    """ Keeps track of the state of the resolution of a variable with a given name
+
+    - Manages the states (unresolved vs. error / success)
+    - Keeps track of what ended up setting the value
+    - Keeps track of what caused the resolvable to enter existence (it's cause)
+    """
+
+    UNRESOLVED = 'unresolved'
+    ERROR = 'error'
+    RESOLVED = 'resolved'
+
+    def __init__(self, name, cause):
+        self._state = Resolvable.UNRESOLVED
+        self.name = name
+        self.cause = cause
+        self.error = None
+        self.setter = None
+        self._value = None
+
+    @property
+    def is_finalized(self):
+        return self._state != Resolvable.UNRESOLVED
+
+    @property
+    def is_error(self):
+        return self._state == Resolvable.ERROR
+
+    @property
+    def is_resolved(self):
+        return self._state == Resolvable.RESOLVED
+
+    def finalize_error(self, error):
+        assert self._state == Resolvable.UNRESOLVED
+        self._state = 'error'
+        self.error = error
+
+    def finalize_value(self, value: str, setter: Setter):
+        assert self._state == 'unresolved'
+        self._state = 'resolved'
+        self._value = value
+        self.setter = setter
+
+    @property
+    def value(self):
+        assert self._state == 'resolved', "is_resolved() must return true for this function to " \
+            "be called. State: {}, is_resolved: {}".format(self._state, self.is_resolved)
+
+        return self._value
+
+    def __str__(self):
+        return "<Resolvable name: {}, state: {}>".format(self.name, self._state)
+
+
+def yield_implicit_validates(target):
```

Comment:
Moving to be a member function but keeping the same name. Not having to hand-aggregate the recursion / trampoline up lists because yield will do it for us is really handy